Frends.AmazonS3.DownloadObject/Frends.AmazonS3.DownloadObject/DownloadObject.cs

```diff
@@ -11,216 +11,179 @@
-    {
-        string responseBody;
-        using (var reader = new StreamReader(responseStream)) responseBody = await reader.ReadToEndAsync();
-
-        if (File.Exists(fullPath))
-        {
-            var file = new FileInfo(fullPath);
+                if (!fileExists || (fileExists && input.DestinationFileExistsAction.Equals(DestinationFileExistsAction.Overwrite) && !FileLocked(input.FileLockedRetries, fullPath, cancellationToken)))
+                {
+                    File.WriteAllText(fullPath, responseBody);
 
-            while (IsFileLocked(file)) Thread.Sleep(1000);
-            File.Delete(fullPath);
-            File.WriteAllText(fullPath, responseBody);
-            return $@"Download with overwrite complete: {fullPath}";
+                    if (input.DeleteSourceObject)
+                        sourceDeleted = await DeleteSourceFile(amazonS3Client, input.BucketName, fileObject.Key, cancellationToken);
+                }
+                else
+                    throw new Exception($"WriteToFile error: An unexpected error.");
+
+                if (File.Exists(fullPath))
+                    return new SingleResultObject() { FullPath = fullPath, ObjectName = file, Overwritten = input.DestinationFileExistsAction.Equals(DestinationFileExistsAction.Overwrite), SourceDeleted = sourceDeleted };
+                else
+                    throw new Exception("WriteToFile error: An unexpected error.");
+            }
+            throw new Exception("WriteToFile error: An unexpected error.");
         }
-        else
+        catch (Exception ex)
         {
-            if (!Directory.Exists(input.DestinationDirectory)) Directory.CreateDirectory(input.DestinationDirectory);
-            File.WriteAllText(fullPath, responseBody);
-            return $@"Download complete: {fullPath}";
+            throw new Exception($"WriteToFile error: {ex}");
         }
     }
 
-    private static async Task<string> WriteToFile(Input input, S3Object fileObject, AmazonS3Client s3Client, string fullPath)
+    private static bool FileLocked(int fileLockedRetries, string fullPath, CancellationToken cancellationToken)
     {
-        string responseBody;
-
-        var request = new GetObjectRequest
-        {
-            BucketName = input.BucketName,
-            Key = fileObject.Key
-        };
-
-        using (var response = await s3Client.GetObjectAsync(request))
-        using (var responseStream = response.ResponseStream)
-
-
-        using (var reader = new StreamReader(responseStream)) responseBody = await reader.ReadToEndAsync();
-        if (File.Exists(fullPath))
+        try
         {
-            var file = new FileInfo(fullPath);
+            var retries = 0;
 
-            while (IsFileLocked(file)) Thread.Sleep(1000);
-            File.Delete(fullPath);
-            File.WriteAllText(fullPath, responseBody);
-            return $@"Download with overwrite complete: {fullPath}";
+            while (retries <= fileLockedRetries)
```

Comment:
Please use `for`-iteration instead of `while`-loop, since we have a clear constraint how many times we want to iterate. `While` should be used when it is not clear how many time we want to iterate something.